### ISSUE TITLE
How do I create state nodes that have child nodes with actions?

### ISSUE THREAD

**Issue by FieldMarshallVague** [skclusive/Skclusive.Blazor.Samples#13]
Hi Skclusive,

Sorry if this isn't the right way to ask, but I've been struggling with something.  I need to track an object that has child-objects that have their own behaviours.  Trunck > Branch > Leaf.  They all have their own EditName/EditBranches actions, so I can't use the DTO (snapshot) class as the type in the view layer (razor file)  

So, Tree would have a list of Branches, I have Tree as a child of AppState, which is tracked and works properly (since it's on the root).   But when I get to adding Branch, this is a child of Tree, not of AppState, so my interface is

```
ITree {
  List<IBranch> Branches
}

IBranchActions {
  EditName(string)
}

IBranchSnapshot {
  string Name
}
```

Now, IBranch needs to have an 'EditName' method, which is part of the Actions interface, this is what is implemented on the Proxy, not the snapshot.

Do you have any examples of this nested structure?  I cannot see how to do it and have been circulating for a while now.  I've looked at Todo and Flightfinder, but this does not seem to have anything like this.  Is it even possible?  Will it ever be? Or am I making a mistake?

I thought maybe I just put everything in the root, but you can imagine that would get very messy and tracking the IDs (or whatever) to link them back up would be a nightmare.

I'd really appreciate some advice on this.  Thanks!

**Comment by skclusive** [comment]
hi. thanks for trying out state tree. yes. it is possible to define the above nested structure with actions specific to node types. the nesting can go any level.

i just added sample test for your use case here https://github.com/skclusive/Skclusive.Mobx.StateTree/tree/master/test/StateTree.Tests/Models/Nested 

and the test here https://github.com/skclusive/Skclusive.Mobx.StateTree/blob/master/test/StateTree.Tests/TestNested.cs

```cs

            var root = RootType.Create(new RootSnapshot
            {
                Tree = new TreeSnapshot
                {
                    Branches = new IBranchSnapshot []
                    {
                        new BranchSnapshot { Name = "branch 1" },

                        new BranchSnapshot { Name = "branch 2" }
                    }
                }
            });

            Assert.NotNull(root);

            Assert.NotNull(root.Tree);

            Assert.Equal(2, root.Tree.Branches.Count);

            Assert.Equal("branch 1", root.Tree.Branches[0].Name);

            Assert.Equal("branch 2", root.Tree.Branches[1].Name);

            root.Tree.AddBranch(new BranchSnapshot { Name = "branch 4 (typo)" });

            Assert.Equal(3, root.Tree.Branches.Count);

            Assert.Equal("branch 4 (typo)", root.Tree.Branches[2].Name);

            root.Tree.Branches[2].EditName("branch 3");

            Assert.Equal("branch 3", root.Tree.Branches[2].Name);

```

hope this helps

**Comment by FieldMarshallVague** [comment]
Awesome!  Thanks so much for this.  I'll test it tomorrow and let you know.  Cheers, really helps out.

**Comment by FieldMarshallVague** [comment]
Thanks again, that's working for the behaviours, but now I can't treat the objects as the snapshots/dto types.  If I pass the Tree (for example) to a component, using root.Tree, the type is ITree (behaviours).  How do you access the node as its snapshot type?  Obviously there is some merging going on 'under the hood', but how do I cast, say root.Tree.Branches[0] as the BranchSnapShot type?

**Comment by skclusive** [comment]
glad you got the answers yourself :-) someday the verbosity could be avoided using C# code generator.

i have been using it with pretty complex normalised data structure for my own project. recently tested with MobileBlazorBinding (netstandard2.0) as well.  working well so far.

**Comment by FieldMarshallVague** [comment]
That sounds like a good move forward.  I was wondering how we could avoid so much boilerplate, but all redux-style implementations have them (AFAICT).  These changes have stopped the Redux debugger working for me, though.  It's throwing this error:

`Deserialization of interface types is not supported. Type 'IRoot'`

This happens when I try to step to a previous state, using the back arrow in the redux debugger.  I can see why it's throwing it, but I am not sure what to do about it.  That is, where to put the concrete type (Root).

**Comment by FieldMarshallVague** [comment]
Ah, I fixed that by setting the interface types to map to an array (I forgot it was talking to JS):

`services.TryAddSingletonEnumerable<JsonConverter, JsonTypeConverter<IEnumerable<IBranch>, Branch[]>>();`

And while this doesn't work when I have the shift-alt-D debugger attached (WASM debugger), it doesn't throw an error until I try to step back in the redux history.  

` Specified cast is not valid.`

It doesn't like the casting of IBranch to BranchSnapshot at this point, I think.

**Comment by skclusive** [comment]
you are almost there. we have to only map the snapshot types. not the observable types. as below.

https://github.com/skclusive/Skclusive.Blazor.Samples/blob/master/Skclusive.Blazor.FlightFinder/FlightFinder.State/Extension/FlightFinderExtension.cs#L17

```cs
services.TryAddJsonTypeConverter<IRootSnapshot, RootSnapshot>();
services.TryAddJsonTypeConverter<IBranchSnapshot, BranchSnapshot>();
services.TryAddJsonTypeConverter<ITreeSnapshot, TreeSnapshot>();
```

can you try the above and let me know if that works?

**Comment by skclusive** [comment]
btw i guess you are using the latest version 5.1.2.

**Comment by FieldMarshallVague** [comment]
Ah, yes, I already tried that, sorry, the code above was a desperate attempt to see if something else worked :)  Sorry, that line was confusing, the error was more clear.

Yes, I'm using the latest versions 5.1.2.

Here is the error again, with some unnecessary bits removed:

`
Uncaught (in promise) Error: System.InvalidCastException: Specified cast is not valid.
   at Skclusive.Mobx.StateTree.ComplexType`2[[DataTypes.IBranchSnapshot[]],[Skclusive.Mobx.Observable.IObservableList`2[[Skclusive.Mobx.StateTree.INode, Skclusive.Mobx.StateTree],[Models.IBranch]], Skclusive.Mobx.Observable]].ApplySnapshot(INode node, Object snapshot)
   at Skclusive.Mobx.StateTree.ListType`2[[DataTypes.IBranchSnapshot, DataTypes],[Models.IBranch, State]].ApplySnapshot(INode node, Object snapshot)
`

**Comment by skclusive** [comment]
oh. this one. i used to get this error. then i fixed the definition. could not recollect now. will try the same structure and update you.

**Comment by FieldMarshallVague** [comment]
Hi Skclusive,  did your latest release of the samples have anything in relating to this?

**Comment by skclusive** [comment]
no. but i just created a new branch with related changes for this issue.

https://github.com/skclusive/Skclusive.Blazor.Samples/tree/Issue13

and this commit https://github.com/skclusive/Skclusive.Blazor.Samples/commit/e6dd2890e8781ff6bb05062f29f997e995ce2e64

i am not able to reproduce the issue. i could time travel in redux devtools with nested structure. may be i miss the use case, have a look and let me know your thoughts.

**Comment by FieldMarshallVague** [comment]
Thanks! I'll take a look.

**Comment by FieldMarshallVague** [comment]
OK, I think I see the problem.  I've been using lists on my TreeSnapshot model, so have been declaring the branches as IEnumerable<IBranchSnapshot>.  But this seems to be where the history navigation blows up.  It can't cast from the IEnumerable<IBranchSnapshot> to the IBranchSnapshot[].   

FYI, I've also tried just using 'Branchsnapshot' (not the interface), but this didn't help.

Where would I have to override/explicitly cast the type to be able to use IEnumerable?  I'm using this model in lists elsewhere, so it's handy to define it as the base.

**Comment by skclusive** [comment]
I guess it is not straight forward. I will have to try with enumeration and update you.

**Comment by FieldMarshallVague** [comment]
Ah, thankyou.  That would be great.

**Comment by FieldMarshallVague** [comment]
Just to be 100% clear, I am using `IEnumerable<IBranchsnapshot>` instead of IBranchSnapshot[].  I think you understood this, but didn't want to waste your time :)  

(github hid the generic part before, because of the angle-bracket syntax)

**Comment by FieldMarshallVague** [comment]
Happy new year :)  I hope you had a good time.

I wondered if you had any luck in using an enumerable?

**Comment by skclusive** [comment]
Happy new year :-)

Missed the track. 

could you please share a minimal repo to reproduce? Or a test case in MobxStateTree ?

**Comment by FieldMarshallVague** [comment]
Hi Skclusive,  yes,  I'll try one or the other and let you know. Cheers.
